### PR TITLE
First draft of 21.0.0.2 GA blog post

### LINKED ISSUE #1091: Open Liberty GA release blog post for 21.0.0.2
This issue is to track the tasks involved in creating and publishing the releases notes for an Open Liberty 4-weekly release.

The Open Liberty release notes are presented in three places:
- [Open Liberty blog](https://openliberty.io/blog/) (GA and beta features; [example](https://openliberty.io/blog/2020/05/07/EJB-persistent-timers-20005.html))
- [Red Hat Runtimes release notes for Open LIberty](https://access.redhat.com/documentation/en-us/open_liberty/2020/) (GA features only; [example](https://access.redhat.com/documentation/en-us/open_liberty/2020/html/release_notes_for_open_liberty_20.0.0.5_on_red_hat_openshift_container_platform/features))
- [Red Hat Developer blog](https://developers.redhat.com/blog/) (GA features only; [example](https://developers.redhat.com/blog/2020/05/13/open-liberty-20-0-0-5-brings-updates-to-ejb-persistent-timers-coordination-and-failover-across-members/))

## 1. Open Liberty release blog post

The Open Liberty release blog post is written in asciidoc and pushed to the [OpenLiberty/blogs](https://github.com/openliberty/blogs) repo. The repo README contains [instructions on how to create, build, and review](https://github.com/OpenLiberty/blogs/blob/prod/README.md) the post, including a template file to help you.

Follow the instructions in the README and the template file to complete the following tasks:

- [x] Create a draft release blog post in asciidoc, build to [the draft site](https://draft-openlibertyio.mybluemix.net/blog/), and create a PR to the `staging` branch with a link to the post on [the draft site](https://draft-openlibertyio.mybluemix.net/blog/). Add all the reviewers (see below) to the PR so that they can add all their review feedback and approvals in the PR.
- [x] Get the draft post reviewed by:
  - [x] All the people who contributed the content to the blog post.
  - [x] At least one of the following people: @mbroz2
- [x] Agree with @mbroz2 which feature will lead this blog post, then write a title, slug, summary first paragraph, and SEO front matter appropriately.
- [x] Get the post approved by @mbroz2.
- [x] On release day (usually a Friday) @mbroz2 will publish the post.

## 2. Red Hat Runtimes release notes for Open Liberty

The release notes are automatically built from the asciidoc release blog post. This is done by using the appropriate tags in the asciidoc source and then the Red Hat documentation build includes the tagged sections in the release notes. The tags ensure that information in the blog post that is not relevant to the Red Hat Runtimes release are omitted from the Red Hat Runtimes release notes.

- [x] In the asciidoc source of the draft release blog post, check that the include tags are in the correct places in the asciidoc source file:
  - [ ] `// tag::intro[]` (before the first paragraph)
  - [ ] `// end::intro[]` (after the link to the list of fixed bugs)
  - [ ] `// tag::run[]` (before the Maven coordinates)
  - [ ] `//end::run[]` (after the Dockerfile entry)
  - [ ] `//tag::features[]` (before the first GA feature section)
  - [ ] `//end::features[]` (after the last GA feature section)
- [x] In [the release notes file](https://raw.githubusercontent.com/PurpleLiberty/docs/master/releasenotes/master-remote.adoc) (in a separate repo on GitHub), find the include statement in the **Features** section of the release notes. Edit the file name part of the include statement so that it targets the new Open Liberty release blog post .adoc file. Don't change anything else. Work with the ID team if they contact you to resolve any problems when the Red Hat build is run.

## All done?

Check that you've completed every task above. Select each check box to confirm that you have before closing this issue.

### PR DESCRIPTION
**Pull request by jakub-pomykala**
Draft of GA blog can be found [here](https://draft-openlibertyio.mybluemix.net/blog/2021/02/19/configure-trustedHeaderOrigin-21002.html).

And here is a full page screenshot of it:
image

### CODE REVIEW

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]
+
+//tag::features[]
+
+[#trustedHeader]
+=== HTTP Channel Configuration Improvements
+
+"WebSphere private headers" in the form `$WSXX` are used by WebSphere-aware proxies to provide information about original requests. The values provided by these headers are made available to the application via `ServletRequest` APIs like `getRemoteHost()`. Two HTTP Dispatcher custom properties are provided by Liberty to control over which remote hosts are allowed to send WebSphere private headers: `trustedHeaderOrigin` and `trustedSensitiveHeaderOrigin`. Ideally, these properties should be configured to only trust known proxy servers which will be forwarding requests to the Liberty server.
+
+Previously, `trustedHeaderOrigin` and `trustedSensitiveHeaderOrigin` only accepted `*`, `none`, or a list of full IP addresses (eg. `"127.0.0.1, 192.168.6.6"`). As requested by customers, both properties can now additionally be configured with IP address wildcards and hostnames. As an example, the following list which will be valid for either property:`"localhost, 127.0.0.1, 192.168.*.*, 0:0:0:0:0:ffff:*:*, *.ibm.com"`
```
Comment (wtlucy) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: The example list at the end here isn't displaying with code formatting correctly, but I'm not sure why
`"localhost, 127.0.0.1, 192.168.*.*, 0:0:0:0:0:ffff:*:*, *.ibm.com"`

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3ARELEASE_VERSION+label%3A%22release+bug%22[21.0.0.2].
+// end::intro[]
+
+
+// tag::run[]
+[#run]
+
+== Run your apps using 21.0.0.2
+
+If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Don't change this yet but if it looks good on the beta blog then we might consider moving this section to the bottom??

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+[link=https://stackoverflow.com/tags/open-liberty]
+image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]
+
+//tag::features[]
+
+[#trustedHeader]
+=== HTTP Channel Configuration Improvements
+
+"WebSphere private headers" in the form `$WSXX` are used by WebSphere-aware proxies to provide information about original requests. The values provided by these headers are made available to the application via `ServletRequest` APIs like `getRemoteHost()`. Two HTTP Dispatcher custom properties are provided by Liberty to control over which remote hosts are allowed to send WebSphere private headers: `trustedHeaderOrigin` and `trustedSensitiveHeaderOrigin`. Ideally, these properties should be configured to only trust known proxy servers which will be forwarding requests to the Liberty server.
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]:
```suggestion
"WebSphere private headers" in the form `$WSXX` are used by WebSphere-aware proxies to provide information about original requests. The values provided by these headers are made available to the application via `ServletRequest` APIs such as `getRemoteHost()`. Two HTTP Dispatcher custom properties are provided by Open Liberty to control which remote hosts are allowed to send the `trustedHeaderOrigin` and `trustedSensitiveHeaderOrigin` private headers. Ideally, these dispatcher properties should be configured to only trust known proxy servers which will be forwarding requests to the Open Liberty server.
```

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+
+[#bugs]
+== Significant bugs fixed in this release
+
+
+We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21002+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.2].
+
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/15433[System WABs may come online with the web container after server reports started]
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]:
```suggestion
* link:https://github.com/OpenLiberty/open-liberty/issues/15433[System Web Application Bundles may not be fully provisioned until after the server reports as started]
```

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
++
+When a WAB is included in a Liberty feature (also called a system WAB) it is possible that the server started message and the TCP ports will be opened and listening before the WAB has been fully provisioned with the web container. This is different behaviour to web applications WARs where the server started and TCP ports are delayed until all the applications have started or a timeout has occurred on server start.
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]:
```suggestion
When a Web Application Bundle (WAB) is included in a Liberty feature (also called a system WAB) it is possible that the server started message and the TCP ports will be opened and listening before the WAB has been fully provisioned with the web container. This is different behaviour to web applications WARs where the server started and TCP ports are delayed until all the applications have started or a timeout has occurred on server start.
```

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+This allows the `WABInstaller` to check on the status of the system WABs to make sure the have completed their provisioning to the web container. 
+
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/15698[FeatureUtility not parsing liberty custom environment variables]
++
+When the entry is `/${shared.config.dir}/server.xml”` and the environment variable is `WLP_USER_DIR=/test/` in the `test/config/server.xml` file the user recieves an error stating that "it couldn't find file at /user/test/wlp/usr/shared/config/server.xml" where it should have been looking for `test/config/server.xml`.
+
+This bug causes include tags to not be prased and replaces the environment variables found link:https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_admin_customvars.html[here]
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]:
```suggestion
This bug causes include tags to not be parsed and replaces the environment variables found link:https://openliberty.io/docs/21.0.0.1/reference/config/server-configuration-overview.html#server-env[here] 
```
Check that new link with the developer

Comment (tevans78) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Also, not sure if the whole sentence quite makes sense?? It almost reads like the bug causes environment variables to be replaced .... maybe there should be a _not_ in there somewhere??

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
+at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
+at java.lang.reflect.Method.invoke(Method.java:508)
+at com.ibm.ws.jaxrs20.injection.HttpServletRequestInjectionProxy$1.invoke(HttpServletRequestInjectionProxy.java:58)
+----
+
+This issue was resolved by adding a null checks to these two proxy classes similar to those in other proxy classes.
+
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/14109[Update gRPC dependencies to 1.35]
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Move this GRPC issue to be first

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
++
+Currently Open Liberty uses `grpc-java` version 1.31.1 to provide its `grpc-1.0` and `grpcClient-1.0` features. This issue will update the `grpc-java` dependencies we build on to their latest version which is 1.35.0. The 1.35.0 release we're pulling in provides numerous improvements and bug fixes; for a full list see link:https://github.com/grpc/grpc-java/releases[GRPC Java Release Notes]
+
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/14175[Expression Language 3.0 value lookup performance improvement]
```
Comment (tevans78) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: And this one second

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc (hunk 3 above)
Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Done

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc (hunk 4 above)
Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Done

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc (hunk 5 above)
Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Done

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc (hunk 6 above)
Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: changed to : "This bug caused include tags to not be parsed correctly by not replacing the environment variables found "

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc (hunk 7 above)
Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Done

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc (hunk 8 above)
Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc]: Done

**Review by tevans78 — APPROVED** [review]

**Review by tevans78 — CHANGES_REQUESTED** [review]
Actually, you are right, let's move the try it now section to the bottom but a put a link to it at the top

File: posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+java.lang.ClassCastException: com.example.MyWrappedRequest incompatible with com.ibm.ws.webcontainer40.srt.SRTServletRequest40
+at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.(CacheServletWrapper40.java:57)
+at com.ibm.ws.webcontainer40.servlet.factory.CacheServletWrapperFactory40Impl.createCacheServletWrapper(CacheServletWrapperFactory40Impl.java:30)
+at com.ibm.ws.webcontainer.WebContainer.addToCache(WebContainer.java:1231)
+at com.ibm.ws.webcontainer.extension.DefaultExtensionProcessor.handleRequest(DefaultExtensionProcessor.java:538)
+at com.ibm.ws.webcontainer.filter.WebAppFilterChain.invokeTarget(WebAppFilterChain.java:182)
+----
+
+This bug has been removed by fixing a `ClassCastException` when caching a static servlet wrapper under `servlet-4.0` feature.
+
```
Comment (dmuelle) [inline posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc:129]: don't forget to add an 
`// end::features[]`  tag

Comment (jakub-pomykala) [reply posts/2021-02-19-configure-trustedHeaderOrigin-21002.adoc:129]: Done

**Review by wtlucy — DISMISSED** [review]

**Review by tevans78 — APPROVED** [review]

**Pull request merged**